pym/bob/scm/git.py

```diff
@@ -262,6 +263,10 @@ async def __checkoutTagOnBranch(self, invoker, fetchCmd, switch):
         if await invoker.callCommand(["git", "merge-base", "--is-ancestor",
                                      commit, "remotes/origin/"+self.__branch],
                                      cwd=self.__dir):
+            # already on the configured commit -> do nothing
+            if self.__commit and rev_parse.stdout.rstrip() == self.__commit:
+                invoker.warn("Not updating", self.__dir, ". Already on the configured commit.")
+                return
```

Comment:
I think this could be moved even before the `fetchCmd` above. It doesn't even need a warning because the desired commit is already checked out. Could you also add the case for a tag, e.g. something along these lines?

```python
if self.__commit ...
    ...
else:
    # Convert tag to commit. Beware of annotated commits!
    tag = await invoker.runCommand(["git", "rev-parse", "tags/"+self.__tag+"^0"], stdout=True, cwd=self.__dir)
    if tag.returncode == 0 and tag.stdout == rev_parse.stdout:
        return
```

I'm wondering if this should all be done only if `headValid == True`..?

Reply:
Thanks. I modified the code following you're suggestions. My assumption was that the output of `git rev-parse` did not match the commit/tag if the HEAD is not valid but this might be wrong. Moving behind the `headValid` should be safe.